memory/mergeable_index.go

```diff
@@ -0,0 +1,284 @@
+		} else {
+			merged = append(merged, lookup)
+		}
+	}
+
+	return &MergedIndexLookup{
+		Intersections: merged,
+		Index:         idx,
+	}
+}
+
+func (i *MergeableIndexLookup) Union(lookups ...sql.IndexLookup) sql.IndexLookup {
+	return union(i.Index, i, lookups...)
+}
+
+// Unions the lookups given together, collapsing redundant layers of unions for lookups that have previously been
+// merged. E.g. merging a MergeableIndexLookup with a MergedIndexLookup that has 2 unions will return a
+// MergedIndexLookup with 3 lookups unioned: the left param and the two unioned lookups from the MergedIndexLookup.
+func union(idx ExpressionsIndex, left sql.IndexLookup, lookups ...sql.IndexLookup) sql.IndexLookup {
+	var merged []sql.IndexLookup
+	var allLookups []sql.IndexLookup
+	allLookups = append(allLookups, left)
+	allLookups = append(allLookups, lookups...)
+	for _, lookup := range allLookups {
+		if mil, ok := lookup.(*MergedIndexLookup); ok && len(mil.Unions) > 0 {
+			merged = append(merged, mil.Unions...)
+		} else {
+			merged = append(merged, lookup)
+		}
+	}
+
+	return &MergedIndexLookup{
+		Unions: merged,
+		Index:  idx,
+	}
+}
+
+// MergedIndexLookup is an index lookup that has been merged with another.
+// Exactly one of the Unions or Intersections fields should be set, and correspond to a logical AND or OR operation,
+// respectively.
+type MergedIndexLookup struct {
+	Unions        []sql.IndexLookup
+	Intersections []sql.IndexLookup
+	Index         ExpressionsIndex
+}
+
+var _ sql.Mergeable = (*MergedIndexLookup)(nil)
+var _ sql.SetOperations = (*MergedIndexLookup)(nil)
+var _ memoryIndexLookup = (*MergedIndexLookup)(nil)
+
+func (m *MergedIndexLookup) EvalExpression() sql.Expression {
+	var exprs []sql.Expression
+	if m.Intersections != nil {
+		for _, lookup := range m.Intersections {
+			exprs = append(exprs, lookup.(memoryIndexLookup).EvalExpression())
+		}
+		return and(exprs...)
+	}
+	if m.Unions != nil {
+		for _, lookup := range m.Unions {
+			exprs = append(exprs, lookup.(memoryIndexLookup).EvalExpression())
+		}
+		return or(exprs...)
+	}
+	panic("either Unions or Intersections must be non-nil")
+}
+
+func (m *MergedIndexLookup) Intersection(lookups ...sql.IndexLookup) sql.IndexLookup {
+	return intersection(m.Index, m, lookups...)
+}
+
+func (m *MergedIndexLookup) Union(lookups ...sql.IndexLookup) sql.IndexLookup {
+	return union(m.Index, m, lookups...)
+}
+
+func (m *MergedIndexLookup) Difference(...sql.IndexLookup) sql.IndexLookup {
+	panic("not implemented")
+}
+
+func (m *MergedIndexLookup) IsMergeable(lookup sql.IndexLookup) bool {
+	_, ok := lookup.(MergeableLookup)
+	return ok
+}
+
+func (m *MergedIndexLookup) Values(p sql.Partition) (sql.IndexValueIter, error) {
+	return &dummyIndexValueIter{
+		tbl:             m.Index.MemTable(),
+		partition:       p,
+		matchExpression: m.EvalExpression(),
+	}, nil
+}
+
+func (m *MergedIndexLookup) Indexes() []string {
+	panic("implement me")
```

Comment:
Assuming we aren't implementing these just yet?

Reply:
Never going to be implemented -- this method only matters for a primary index, so that the engine can match columns to indexes. For this synthetic index, that never happens.

Reply:
Ah gotcha. I'd change the panic message then in that case.